### PR TITLE
build: update to jasmine 4.0

### PR DESCRIPTION
**Pull request by crisbeto**
Updates us to version 4.0 of Jasmine and fixes some errors that were the result of us depending upon deprecated APIs. We need to do this both to stay up to date and because it was going to break eventually, because one of the Bazel packages was logging a deprecation warning that version 4.0 was required.

There were also some cases where the state of `ngDevMode` had started leaking out between tests.

**Note:** we also need to update AIO to the new version. I'll do that separately since the non-AIO code changes ended up being more than I expected.

### CODE REVIEW

File: packages/platform-browser/test/testing_public_spec.ts [hunk 1]
```diff
@@ -200,58 +200,58 @@ const bTok = new InjectionToken<string>('b');
         beforeEach(() => {
           TestBed.configureTestingModule(
               {providers: [{provide: FancyService, useValue: new FancyService()}]});
+        });
```
Comment (crisbeto) [inline packages/platform-browser/test/testing_public_spec.ts:203]: This change is interesting: all the tests were defined inside `beforeEach` which meant that they weren't actually running.

**Review by gkalpak — APPROVED** [review]

File: packages/platform-browser/testing/src/matchers.ts [hunk 2]
```diff
@@ -130,22 +131,6 @@ export const expect: <T = any>(actual: T) => NgMatchers<T> = _global.expect;
 };
 
 _global.beforeEach(function() {
-  // Custom handler for Map as we use Jasmine 2.4, and support for maps is not
-  // added until Jasmine 2.6.
-  jasmine.addCustomEqualityTester(function compareMap(actual: any, expected: any): boolean {
```
Comment (devversion) [inline packages/platform-browser/testing/src/matchers.ts:135]: @crisbeto Someone else sent a PR for this, might want to coordinate on that: https://github.com/angular/angular/pull/45521

**Review by devversion — APPROVED** [review]

**Review by JiaLiPassion — APPROVED** [review]
LGTM!

**Review by AndrewKushnir — APPROVED** [review]

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit a248df06824db1c40346b84de07ff905b0d0606f.

**Comment by jessicajaniuk** [comment]
This causes breakages in Google3 and has been rolled back. We'll need a TGP to get this to land safely.

**Comment by AndrewKushnir** [comment]
@jessicajaniuk @crisbeto  I believe the breakages are related to the Maps shim removal, see https://github.com/angular/angular/pull/45521#issuecomment-1092752009.

**Comment by crisbeto** [comment]
I have pushed a change to make this easier to land. Can we give it another shot?

**Comment by jessicajaniuk** [comment]
We'll need a TGP before we can attempt landing this again. I've added `action: presubmit` to indicate that need.

**Comment by crisbeto** [comment]
[Passing TGP](https://fusion2.corp.google.com/presubmit/tap/440566992/OCL:440566992:BASE:440573079:1649513619841:214a789e). Note that I had to re-run once in order to fix a few unrelated flakes. Setting this as ready to merge again.

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 41223a81f2499e765d544122673b89b1ff754e7d.

**Pull request closed without merging**